src/storage-controller/src/lib.rs

```diff
@@ -1164,6 +1168,40 @@ where
         Ok(())
     }
 
+    fn drop_tables(
+        &mut self,
+        identifiers: Vec<GlobalId>,
+        ts: Self::Timestamp,
+    ) -> Result<(), StorageError> {
+        assert!(
+            identifiers
+                .iter()
+                .all(|id| self.collections[id].description.is_table()),
+            "identifiers contain non-tables: {:?}",
+            identifiers
+                .iter()
+                .filter(|id| !self.collections[id].description.is_table())
+                .collect::<Vec<_>>()
+        );
+        let drop_notif = self
+            .persist_table_worker
+            .drop_handles(identifiers.clone(), ts);
+        let (tx, _) = tokio::sync::broadcast::channel(1);
+        for id in &identifiers {
+            let mut rx = tx.subscribe();
+            let drop_notif = async move {
+                let _ = rx.recv();
+            }
+            .boxed();
+            self.pending_table_handle_drops.insert(*id, drop_notif);
+        }
+        mz_ore::task::spawn(|| "table-cleanup".to_string(), async move {
+            drop_notif.await;
+            let _ = tx.send(());
+        });
```

Comment:
This is pretty gross, so all suggestions are welcome. If you want to look at a cleaner alternative that I tried, but didn't work, you can see this commit: https://github.com/MaterializeInc/materialize/pull/26291/commits/01bbb0224430f4e001b883e05dd5804177e696ee. 

The issue is that we cannot finalize a shard until it has both been dropped from the persist table worker and it has been downgraded by the persist read worker. We need to wait for the shard drop because part of dropping a shard from persist involves advancing the upper of that shard which would fail if the shard has been finalized. I'm less clear about why we need to wait for the downgrade, but if we don't wait the following panic happens:

1. Create table.
2. Subscribe to table.
3. Drop table.
4. Drop shard via persist table worker.
5. Finalize shard.
6. Subscribe is cancelled and attempts to downgrade it's read hold on the table.
7. Subscribe panics because it cannot find the table.

Another approach would be to plumb the `drop_notif` via a new field in `PendingCompactionCommand`. That would require optionally passing a timestamp to `Controller::update_read_capabilities` and `Controller::set_read_policy`, which felt a bit more invasive to me.